webhook/resource_queue.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/frostbyte73/core"
+	"github.com/gammazero/deque"
+	"github.com/livekit/protocol/livekit"
+)
+
+var (
+	errQueueBackedUp = errors.New("queue is backed up")
+	errQueueFull     = errors.New("queue is full")
+)
+
+type item struct {
+	ctx      context.Context
+	queuedAt time.Time
+	event    *livekit.WebhookEvent
+}
+
+type resourceQueueParams struct {
+	MaxAge   time.Duration
+	MaxDepth int
+
+	Poster poster
+}
+
+type resourceQueue struct {
+	params resourceQueueParams
+
+	mu    sync.Mutex
+	items deque.Deque[*item]
+	cond  *sync.Cond
+
+	closed core.Fuse
+	drain  atomic.Bool
+}
+
+func newResourceQueue(params resourceQueueParams) *resourceQueue {
+	r := &resourceQueue{
+		params: params,
+	}
+	r.items.SetBaseCap(int(min(params.MaxDepth, 16)))
+	r.cond = sync.NewCond(&r.mu)
+
+	go r.worker()
+	return r
+}
+
+func (r *resourceQueue) Stop(force bool) {
+	r.drain.Store(!force)
+	r.closed.Break()
+	r.cond.Broadcast()
+}
+
+func (r *resourceQueue) Enqueue(ctx context.Context, whEvent *livekit.WebhookEvent) error {
+	return r.EnqueueAt(ctx, time.Now(), whEvent)
+}
+
+func (r *resourceQueue) EnqueueAt(ctx context.Context, at time.Time, whEvent *livekit.WebhookEvent) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.items.Len() >= r.params.MaxDepth {
+		return errQueueFull
+	}
+
+	if r.items.Len() > 0 {
+		front := r.items.Front()
+		if time.Since(front.queuedAt) > r.params.MaxAge {
```

Comment:
we should prune expired items in the worker loop rather than rejecting new submissions

Reply:
I was debating between the two. Cool. Will move it to the processing queue.

Reply:
Implemented changes in these commits
- https://github.com/livekit/protocol/pull/1000/commits/4b5ccabd128bc4af664de3fa86a8bddbabb66ada
- https://github.com/livekit/protocol/pull/1000/commits/ac35fa74c9c2fe9d9e52a2e45dae2f8243ca0ac9
- https://github.com/livekit/protocol/pull/1000/commits/7b248aa039ded5440acf4e1152f13f296346f24d
(sorry, had some test errors, so multiple commits).

Please have a look when you get a chance. If these look good, I will switch the default notifier to this before merging.